Use TaskGroups instead of array of tasks

Frizlab opened the pull request:
Rationale for this comes from this thread https://forums.swift.org/t/taskgroup-vs-an-array-of-tasks/53931
Apparently launching a lot of Tasks just like that is not very good, and we should use task groups instead.

I did _not_ test the performance implication of this change, though. Maybe it’s actually worse because of the additional sort needed. In theory it should be better though.

Frizlab committed:
Use TaskGroups instead of array of tasks

AndrewBarba commented:
Instead of sorting does it make sense to do something like:

```swift
var results = Array<T?>(repeating: nil, count: self.count)
...
for try await result in group {
    results[result.offset] = result.value
}
return results as! [T]
```

Frizlab commented:
@AndrewBarba I’d say yes. And it’s probably faster too. I’ll push your alternative.

Frizlab committed:
Reimplement task group using an array to collect the values

I’m not sure we gain a lot. We’d have to actually measure this solution and the one in the previous commit.

Frizlab commented:
Done. However, `count` is not available in a generic `Sequence`, so I have to grow the result array as I retrieve the results. I’m not sure we actually gain anything compared to the sort solution. We should measure both solution to find the fastest.
Previous implementation is still available here 1105df4.

tachyonics commented:
I don't think it is necessary to have `count` available on the input Sequence. We are iterating through the sequence to add tasks to the group so we *know* how many tasks we have added to the group and therefore how many we are expecting back-

https://github.com/tachyonics/CollectionConcurrencyKit/blob/main/Sources/CollectionConcurrencyKit.swift#L123

Frizlab committed:
Use a better method to retrieve the resulting arrays

Frizlab commented:
@tachyonics You’re not wrong. I have pushed your method.

johnrogers commented:
Fully support these changes. Not only will it make the functions more performant, but it will ensure each function  participates in `Task` cancellation inheritance correctly.
